### DISCUSSION TITLE
Why Jumpy?

### DISCUSSION THREAD

**Discussion by cemlyn007** [google/brax#225] [Q&A] [answered]
Wondering what was the reason for creating Brax to allow NumPy and JAX via jumpy?

Why not just use JAX completely otherwise you have to add a method to jumpy that exists in NumPy and JAX? Then things would be simpler and easier to test (for me at least)?

Maybe I am the only one missing something here. 😅

**Comment by cemlyn007** [comment]
For example what should be the expected outcome of:
```
import brax
import brax.jumpy as jp
env = brax.envs.create('ant')
state = env.reset(jp.random_prngkey(0))
state = env.step(state, jp.random_uniform(jp.random_prngkey(0), (env.action_size,), -1., 1.))
```
When using Brax 0.0.15. Should I expect an error?

**Comment by erikfrey** [comment, marked as answer]
Jumpy allows us to operate Brax in a sort of "immediate mode", where things like physics steps can be run without any jitting overhead - this lowers the time to run a single physics step from tens of seconds to less than a second.  This is very useful when debugging environments or the underlying engine, where you want to make lots of small changes and review each change quickly.

Since we wrote Jumpy, JAX has improved in its ability to be run without jax.jit and return quickly.  So one day we may review this decision and it's time to retire Jumpy... but not today.

**Reply by cemlyn007** [reply]
Thank you for your detailed response. Have a good weekend!

**Reply by jongyaoY** [reply]
> Jumpy allows us to operate Brax in a sort of "immediate mode", where things like physics steps can be run without any jitting overhead - this lowers the time to run a single physics step from tens of seconds to less than a second. This is very useful when debugging environments or the underlying engine, where you want to make lots of small changes and review each change quickly.
> 
> Since we wrote Jumpy, JAX has improved in its ability to be run without jax.jit and return quickly. So one day we may review this decision and it's time to retire Jumpy... but not today.

I have seen the latest brax is not dependent on jumpy anymore, does that mean "its ability to be run without jax.jit and return quickly" is realized in some other better way? or jitting becomes mandatory now?
